src/ngx_http_lua_fd.c

```diff
@@ -0,0 +1,229 @@
+
+        /* invalidate connection object */
+        u->conn->fd = -1;
+        ngx_free_connection(u->conn);
+
+        u->conn = NULL;
+    }
+    ngx_free(u);
+    co_ctx->data = NULL;
+}
+
+
+static int
+ngx_http_lua_fd_wait(lua_State *L)
+{
+    ngx_http_request_t    *r;
+    ngx_http_lua_ctx_t    *ctx;
+    ngx_http_lua_co_ctx_t *co_ctx;
+    ngx_http_lua_udata_t  *u;
+
+    ngx_socket_t fd = luaL_optint(L, 1, -1); /* -1 is invalid fd */
+    const char *events = luaL_optstring(L, 2, "");
+    int poll_mask = 0;
+    double timeout = luaL_optnumber(L, 3, HUGE_VAL); /* default to infinite timeout */
+    while (*events) {
+        if (*events == 'r') {
+            poll_mask |= POLLIN;
+        } else if (*events == 'w') {
+            poll_mask |= POLLOUT;
+        }
+        events++;
+    }
+    if ((fd < 0 || !poll_mask) && timeout == HUGE_VAL) {
+        return luaL_error(L, "must provide a valid file descriptor or timeout");
+    }
+    if ((r = ngx_http_lua_get_req(L)) == NULL) {
+        return luaL_error(L, "no request found");
+    }
+    if ((ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module)) == NULL) {
+        return luaL_error(L, "no request ctx found");
+    }
+    if ((co_ctx = ctx->cur_co_ctx) == NULL) {
+        return luaL_error(L, "no co ctx found");
+    }
+    if ((u = ngx_alloc(sizeof(ngx_http_lua_udata_t), r->connection->log)) == NULL) {
+        return luaL_error(L, "no mem");
+    }
+    /* always create a connection object (even if fd is < 0) */
+    if ((u->conn = ngx_get_connection(fd, r->connection->log)) == NULL) {
+        ngx_free(u);
+        return luaL_error(L, "unable to get nginx connection");
+    }
+    ngx_http_lua_cleanup_pending_operation(co_ctx);
+
+    u->conn->data = u;
+    u->conn->read->handler = ngx_http_lua_fd_rev_handler;
+    u->conn->read->log = u->conn->log;
+    u->conn->write->handler = ngx_http_lua_fd_rev_handler;
+    u->conn->write->log = u->conn->log;
+    u->conn->number = ngx_atomic_fetch_add(ngx_connection_counter, 1);
+
+    if (fd >= 0) {
+        if ((poll_mask & POLLIN) && ngx_handle_read_event(u->conn->read, NGX_LEVEL_EVENT) != NGX_OK) {
+            ngx_free_connection(u->conn);
+            ngx_free(u);
+            return luaL_error(L, "unable to add to nginx main loop");
+        }
+        if ((poll_mask & POLLOUT) && ngx_handle_write_event(u->conn->write, NGX_LEVEL_EVENT) != NGX_OK) {
+            if (poll_mask & POLLIN) {
+                ngx_del_event(u->conn->read, NGX_READ_EVENT, 0);
+            }
+            ngx_free_connection(u->conn);
+            ngx_free(u);
+            return luaL_error(L, "unable to add to nginx main loop");
+        }
+    }
+
+    u->request = r;
+
+    u->co_ctx = co_ctx;
+    co_ctx->cleanup = (ngx_http_cleanup_pt)&ngx_http_lua_fd_cleanup;
+    co_ctx->data = u;
+
+    if (timeout != HUGE_VAL) { /* no point adding an infinite timeout */
+        ngx_add_timer(u->conn->read, (ngx_msec_t)(timeout * 1000));
```

Comment:
> when u->conn is actively waiting for the next read event.

But can you still use `conn->read`'s timer object when the connection is only waiting for a write event?
Reading through the source it seems okay to do.

Otherwise I'll need to add this patch:
```diff
diff --git a/src/ngx_http_lua_fd.c b/src/ngx_http_lua_fd.c
index d5dc6db..f46aa36 100644
--- a/src/ngx_http_lua_fd.c
+++ b/src/ngx_http_lua_fd.c
@@ -99,6 +99,9 @@ ngx_http_lua_fd_cleanup(ngx_http_lua_co_ctx_t *co_ctx)
         if (u->conn->read->timer_set) {
             ngx_del_timer(u->conn->read);
         }
+        if (u->conn->write->timer_set) {
+            ngx_del_timer(u->conn->read);
+        }
 
         if (u->conn->fd != -1) {
             /* remove from mainloop; do not pass CLOSE_SOCKET flag */
@@ -207,7 +210,7 @@ ngx_http_lua_fd_wait(lua_State *L)
     co_ctx->data = u;
 
     if (timeout != HUGE_VAL) { /* no point adding an infinite timeout */
-        ngx_add_timer(u->conn->read, (ngx_msec_t)(timeout * 1000));
+        ngx_add_timer((poll_mask & POLLIN) ? u->conn->read : u->conn->write, (ngx_msec_t)(timeout * 1000));
     }
 
     /* make sure nginx knows what to do next */
```